File: backend/services/math_agent.py
"""
Enhanced Math Routing Agent with Agentic-RAG architecture.
Includes intelligent routing between knowledge base and web search with MCP.
"""

import asyncio
import time
import sys
sys.path.append('/backend/services')
from typing import Dict, Any, Optional
from datetime import datetime
try:
    from .math_solver import AdvancedMathSolver
except ImportError:
    from math_solver import AdvancedMathSolver

class MathRoutingAgent:
    def __init__(self, storage, guardrails, vector_db, web_search, feedback_system):
        self.storage = storage
        self.guardrails = guardrails
        self.vector_db = vector_db
        self.web_search = web_search
        self.feedback_system = feedback_system
        self.similarity_threshold = 0.85  # Higher threshold - only use exact matches
        self.math_solver = AdvancedMathSolver()  # Real mathematical solver
        
    def _categorize_problem(self, problem: str) -> str:
        """Intelligently categorize the mathematical problem."""
        problem_lower = problem.lower()
        
        # Calculus indicators
        if any(keyword in problem_lower for keyword in 
               ["derivative", "differentiate", "integrate", "integration", "limit", "d/dx", "∫"]):
            return "calculus"
        
        # Algebra indicators
        elif any(keyword in problem_lower for keyword in 
                ["equation", "solve", "factor", "quadratic", "linear", "polynomial", "system"]):
            return "algebra"
        
        # Geometry indicators
        elif any(keyword in problem_lower for keyword in 
                ["triangle", "circle", "area", "volume", "perimeter", "angle", "coordinate"]):
            return "geometry"
        
        # Statistics indicators
        elif any(keyword in problem_lower for keyword in 
                ["mean", "median", "mode", "standard deviation", "variance", "probability"]):
            return "statistics"
        
        # Trigonometry indicators
        elif any(keyword in problem_lower for keyword in 
                ["sin", "cos", "tan", "trigonometric", "radian", "degree"]):
            return "trigonometry"
        
        else:
            return "general"
    
    async def solve_problem(
        self, 
        problem: str, 
        show_steps: bool = True, 
        include_explanations: bool = True
    ) -> Dict[str, Any]:
        """
        Enhanced problem-solving pipeline with intelligent routing and categorization.
        """
        start_time = time.time()
        
        # Step 1: Apply AI Guardrails
        validation = await self.guardrails.validate_mathematical_content(problem)
        if not validation["is_valid"]:
            raise ValueError(validation.get("reason", "Invalid mathematical content"))
        
        # Categorize the problem for better routing
        problem_category = self._categorize_problem(problem)
        
        # Log activity with category
        await self.storage.create_activity(
            action="Problem submitted",
            source="user_input",
            details=f"Category: {problem_category}, Problem: {problem[:100]}"
        )
        
        try:
            # Step 2: Search knowledge base first (Vector DB)
            similar_problems = await self.vector_db.search_similar_problems(problem, limit=3)
            
            # Step 2: Use our advanced math solver to actually solve the problem
            solution = self.math_solver.solve_mathematical_problem(problem)
            
            # Determine difficulty based on problem complexity
            difficulty = self._determine_difficulty(problem, problem_category)
            
            # Store the generated solution
            problem_record = await self.storage.create_math_problem(
                problem=problem,
                solution=solution,
                category=problem_category,
                difficulty=difficulty,
                source="math_solver"
            )
            
            await self.storage.create_activity(
                action="Solution generated",
                source="math_solver",
                details=f"Category: {problem_category}, Difficulty: {difficulty}"
            )
            
            return {
                "problem": problem,
                "solution": solution,
                "source": "math_solver",
                "response_time": (time.time() - start_time) * 1000,
                "category": problem_category,
                "problem_id": problem_record["id"],
                "confidence_score": 0.95,  # High confidence in our solver
                "difficulty": difficulty
            }
            
            # Step 3: If solver has issues, fallback to web search with MCP
            # search_results = await self.web_search.search_mathematical_content(problem)
            # web_solution = await self.web_search.generate_solution_from_search(problem, search_results)
            
            # For now, return a helpful message if our solver fails
            web_solution = {
                "steps": [
                    {
                        "step": 1,
                        "title": "Problem Analysis",
                        "content": f"Analyzing: {problem}",
                        "explanation": "Please provide more specific mathematical notation for better analysis"
                    }
                ],
                "final_answer": "Please rephrase your mathematical problem with clear notation"
            }
            
            # Store fallback solution
            problem_record = await self.storage.create_math_problem(
                problem=problem,
                solution=web_solution,
                category=problem_category,
                difficulty="intermediate", 
                source="fallback"
            )
            
            await self.storage.create_activity(
                action="Fallback solution",
                source="fallback",
                details="Solver needs more specific notation"
            )
            
            return {
                "problem": problem,
                "solution": web_solution,
                "source": "fallback",
                "response_time": (time.time() - start_time) * 1000,
                "category": problem_category,
                "problem_id": problem_record["id"],
                "confidence_score": 0.6,
                "difficulty": "intermediate"
            }
            
        except Exception as error:
            await self.storage.create_activity(
                action="Solution failed",
                source="error",
                details=str(error)
            )
            raise Exception("Failed to solve the mathematical problem. Please try again or rephrase your question.")
    
    def _determine_difficulty(self, problem: str, category: str) -> str:
        """Determine problem difficulty based on content and category."""
        problem_lower = problem.lower()
        
        # Advanced indicators
        if any(keyword in problem_lower for keyword in 
               ["integration by parts", "complex", "limit", "infinite", "derivative of derivative", "system"]):
            return "hard"
        
        # Basic indicators  
        elif any(keyword in problem_lower for keyword in 
                ["simple", "basic", "easy", "find x", "solve for", "calculate"]):
            return "basic"
        
        # Category-based difficulty
        elif category in ["calculus", "trigonometry"]:
            return "intermediate"
        elif category in ["algebra", "geometry"]:
            return "basic"
        
        return "intermediate"
    
    async def submit_feedback(
        self,
        problem_id: str,
        accuracy_rating: int,
        clarity_rating: str,
        comments: Optional[str] = None,
        is_helpful: bool = True
    ):
        """
        Enhanced feedback system with DSPy integration.
        """
        # Store feedback
        await self.storage.create_feedback(
            problem_id=problem_id,
            accuracy_rating=accuracy_rating,
            clarity_rating=clarity_rating,
            comments=comments,
            is_helpful=is_helpful
        )
        
        # Process feedback with DSPy for learning
        await self.feedback_system.process_feedback(
            problem_id=problem_id,
            accuracy_rating=accuracy_rating,
            clarity_rating=clarity_rating,
            comments=comments
        )
        
        await self.storage.create_activity(
            action="Feedback received",
            source="user_feedback",
            details=f"Rating: {accuracy_rating}/5, Clarity: {clarity_rating}"
        )

File: server/services/mathAgent.ts
import { vectorDB } from './vectorDB';
import { webSearch } from './webSearch';
import { guardRails } from './guardrails';
import { storage } from '../storage';

export interface MathSolutionRequest {
  problem: string;
  showSteps: boolean;
  includeExplanations: boolean;
}

export interface MathSolutionResponse {
  problem: string;
  solution: any;
  source: 'knowledge_base' | 'web_search';
  responseTime: number;
  category?: string;
  problemId: string;
}

export class MathRoutingAgent {
  private readonly similarityThreshold = 0.7;

  async solveProblem(request: MathSolutionRequest): Promise<MathSolutionResponse> {
    const startTime = Date.now();

    // Step 1: Apply guardrails
    const validation = await guardRails.validateMathematicalContent(request.problem);
    if (!validation.isValid) {
      throw new Error(validation.reason || "Invalid mathematical content");
    }

    // Log activity
    await storage.createActivity({
      action: "Problem submitted",
      source: "user_input",
      details: request.problem.substring(0, 100)
    });

    try {
      // Step 2: Search knowledge base first
      const similarProblems = await vectorDB.searchSimilarProblems(request.problem, 3);
      
      if (similarProblems.length > 0 && similarProblems[0].similarity > this.similarityThreshold) {
        // Found similar problem in knowledge base
        const bestMatch = similarProblems[0];
        
        const problemRecord = await storage.createMathProblem({
          problem: request.problem,
          solution: bestMatch.solution,
          category: bestMatch.category,
          difficulty: "intermediate",
          source: "knowledge_base"
        });

        await storage.createActivity({
          action: "Solution found",
          source: "knowledge_base",
          details: `Similarity: ${bestMatch.similarity.toFixed(2)}`
        });

        return {
          problem: request.problem,
          solution: bestMatch.solution,
          source: 'knowledge_base',
          responseTime: Date.now() - startTime,
          category: bestMatch.category,
          problemId: problemRecord.id
        };
      }

      // Step 3: Fallback to web search
      const searchResults = await webSearch.searchMathematicalContent(request.problem);
      const webSolution = await webSearch.generateSolutionFromSearch(request.problem, searchResults);

      const problemRecord = await storage.createMathProblem({
        problem: request.problem,
        solution: webSolution,
        category: "advanced",
        difficulty: "advanced",
        source: "web_search"
      });

      await storage.createActivity({
        action: "Solution found",
        source: "web_search",
        details: `Sources: ${searchResults.length} found`
      });

      return {
        problem: request.problem,
        solution: webSolution,
        source: 'web_search',
        responseTime: Date.now() - startTime,
        category: "advanced",
        problemId: problemRecord.id
      };

    } catch (error) {
      await storage.createActivity({
        action: "Solution failed",
        source: "error",
        details: (error as Error).message
      });
      
      throw new Error("Failed to solve the mathematical problem. Please try again or rephrase your question.");
    }
  }

  async submitFeedback(problemId: string, feedback: {
    accuracyRating: number;
    clarityRating: string;
    comments?: string;
    isHelpful: boolean;
  }) {
    await storage.createFeedback({
      problemId,
      ...feedback
    });

    await storage.createActivity({
      action: "Feedback received",
      source: "user_feedback",
      details: `Rating: ${feedback.accuracyRating}/5, Clarity: ${feedback.clarityRating}`
    });
  }
}

export const mathAgent = new MathRoutingAgent();
